lib/compress/zstd_opt.c

```diff
@@ -764,6 +764,180 @@ FORCE_INLINE_TEMPLATE U32 ZSTD_BtGetAllMatches (
+
+    /* Setting match end position to MAX to ensure we never use an LDM during this block */
+    if (ldmSeqStore->pos >= ldmSeqStore->size) {
+        *matchStartPosInBlock = UINT_MAX;
+        *matchEndPosInBlock = UINT_MAX;
+        return;
+    }
+    /* Calculate appropriate bytes left in matchLength and litLength after adjusting
+       based on ldmSeqStore->posInSequence */
+    currSeq = ldmSeqStore->seq[ldmSeqStore->pos];
+    assert(ldmSeqStore->posInSequence <= currSeq.litLength + currSeq.matchLength);
+    currBlockEndPos = currPosInBlock + blockBytesRemaining;
+    literalsBytesRemaining = (ldmSeqStore->posInSequence < currSeq.litLength) ?
+            currSeq.litLength - (U32)ldmSeqStore->posInSequence :
+            0;
+    matchBytesRemaining = (literalsBytesRemaining == 0) ?
+            currSeq.matchLength - ((U32)ldmSeqStore->posInSequence - currSeq.litLength) :
+            currSeq.matchLength;
+
+    /* If there are more literal bytes than bytes remaining in block, no ldm is possible */
+    if (literalsBytesRemaining >= blockBytesRemaining) {
+        *matchStartPosInBlock = UINT_MAX;
+        *matchEndPosInBlock = UINT_MAX;
+        ZSTD_opt_skipBytesInLdmSeqStore(ldmSeqStore, blockBytesRemaining);
+        return;
+    }
+
+    /* Matches may be < MINMATCH by this process. In that case, we will reject them
+       when we are deciding whether or not to add the ldm */
+    *matchStartPosInBlock = currPosInBlock + literalsBytesRemaining;
+    *matchEndPosInBlock = *matchStartPosInBlock + matchBytesRemaining;
+    *matchOffset = currSeq.offset;
+
+    if (*matchEndPosInBlock > currBlockEndPos) {
+        /* Match ends after the block ends, we can't use the whole match */
+        *matchEndPosInBlock = currBlockEndPos;
+        ZSTD_opt_skipBytesInLdmSeqStore(ldmSeqStore, currBlockEndPos - currPosInBlock);
+    } else {
+        /* If we can use the whole match point the ldmSeqStore at the next match */
+        ldmSeqStore->posInSequence = 0;
+        ldmSeqStore->pos++;
+    }
+    DEBUGLOG(6, "ZSTD_opt_getNextLdmAndUpdateSeqStore(): got an ldm that beginning at pos: %u, end at pos: %u, with offset: %u",
+             *matchStartPosInBlock, *matchEndPosInBlock, *matchOffset);
+}
+
+/* ZSTD_opt_maybeAddLdm():
+ * Adds a match if it's long enough, based on it's 'matchStartPosInBlock'
+ * and 'matchEndPosInBlock', into 'matches'. Maintains the correct ordering of 'matches'
+ */
+static void ZSTD_opt_maybeAddLdm(ZSTD_match_t* matches, U32* nbMatches,
+                                 U32 matchStartPosInBlock, U32 matchEndPosInBlock,
+                                 U32 matchOffset, U32 currPosInBlock) {
+    U32 posDiff = currPosInBlock - matchStartPosInBlock;
+    /* Note: ZSTD_match_t actually contains offCode and matchLength (before subtracting MINMATCH) */
+    U32 candidateMatchLength = matchEndPosInBlock - matchStartPosInBlock - posDiff;
+    U32 candidateOffCode = matchOffset + posDiff + ZSTD_REP_MOVE;
```

Comment:
Just realized that I shouldn't be adding `posDiff` here, but it doesn't affect this commit since `posDiff` is guaranteed to be 0 if we proceed.

So after fixing that bug, enabling the functionality to add "extra" LDMs is a one-liner. I.e. for any ldm with `matchLength > 3`, insert all intermediate matches, decrementing `matchLength` until `matchLength == 3`. We simply remove the check for `posDiff > 0` in the condition below. Based on some brief tests just now, it does bring some _slight_ compression ratio benefits on extreme cases, at the cost of severe compression speed reduction.

**Compressing two linux kernels**
Current commit - Ratio: 6.14%, Time: 10m:46s
Adding extra LDMs - Ratio: 6.12% Time: 15m:39s

**Compressing five linux kernels**
Current commit - Ratio: 3.36%, Time: 10m:46s
Adding extra LDMs - Ratio: 3.33% Time: 36m:08s

Reply:
Thanks for these early results !
The compression ratio benefit is in line with expectation,
but the severe speed reduction is not. And is a concern.

It will be interesting to understand why that happens.
Presuming the current strategy of only presenting ldm matchLen > targetLen,
the parser is expected to accept such large matches point blank,
without even considering any further update up to the end of the long match.

Reply:
I guess that part of the reason may be that for `-22`, the `targetLength = 999` - in this case it's pretty likely that we find a bunch of LDMs less than `999`. 

So if our LDM had length 999, we'd potentially insert and process up to a maximum of 996 additional matches over the course of the next 996 positions, and have to perform all the operations that exist in the loops that exist in the lines of code at `zstd_opt.c:1172 - 1197` calculating the relative cost and whatnot, which probably doesn't come for free.

Other than that, removing the `posDiff > 0` line of code also will use up a few more operations in the code that I added, even when rejecting the new partial LDMs, but I don't think that should be too significant since it's mostly a couple of compare operations.

Reply:
Trying `-22 --long` on `silesia.tar` : 
```
../lib/compress/zstd_ldm.c: ZSTD_ldm_adjustParameters
../lib/compress/zstd_ldm.c: minMatchLength : 999 ;  hashRateLog : 7  ; bucketSizeLog : 3
```

The ldm should only deliver matches which are >= 999 bytes.

That being said, if a ldm match can be started mid-flight, it can result in a candidate which has a length < 999.

It could be interesting to trace a few statistics on this topic to visualize better.

Reply:
This speed regression wasn't making sense to me so I tried reproducing it today and couldn't - at which point I realized that in my quick check last Friday I must have ran it without `-T0`.

Though, that does once again bring up the separate issue of the rules for triggering multithreaded mode automatically, since this seems like a case where we'd want multithreaded mode to get triggered.